src/core/heavyGraph.js

```diff
@@ -0,0 +1,122 @@
+// @flow
+import {
+  type Edge,
+  type Node,
+  type EdgeAddressT,
+  type NodeAddressT,
+  Graph,
+} from "./graph";
+import {
+  type NodeWeight,
+  type EdgeWeight,
+  type WeightsComparison,
+} from "./weights";
+
+/**
+ * An entry in the array returned by AddressModule.toParts()
+ */
+export type AddressPart = string;
+
+/**
+ * the core structure in the Node weight-resolution tree
+ */
+export type NodeWeightBranch = {
+  +weight?: NodeWeight,
+  +children: Map<AddressPart, NodeWeightBranch>,
+};
+
+/**
+ * the core structure in the Edge weight-resolution tree
+ */
+export type EdgeWeightBranch = {
+  +weight?: EdgeWeight,
+  +children: Map<AddressPart, EdgeWeightBranch>,
+};
```

Comment:
Have you taken a look at the [Trie](https://github.com/sourcecred/sourcecred/blob/master/src/core/trie.js) module? We should be able to re-use a Trie rather than implementing any tree logic here.